### ISSUE TITLE
PackageReferences: Dependency Management for lib authors

### ISSUE THREAD

**Issue by rcollina** [dotnet/runtime#50042]
Hi!
I've been working on a set of internal libraries for our company, they mostly build upon `M.E.DI`, `M.E.L`, etc.

**Scenario**
Let's say I have a library named `MyCompany.Microsoft.Extensions.Hosting`. 
This library depends on `Microsoft.Extensions.Hosting` version 3.1.3.

I've been trying to figure out how to properly `<PackageReference />` `Microsoft.Extensions.Hosting` to prevent conflicts with other libraries and applications depending on the same Extensions package. 
I'd also like to know if I need to repackage my own library each time a new Patch version is out (3.1.4, for instance), or if there's a smart way to avoid this.

`M.E.H` is an example, I think the same holds for any third-party dependency really.

I could use some guidance here, as getting it wrong would imply several downstream issues for our internal customers.

**What I've tried**
In `MyCompany.Microsoft.Extensions.Hosting.csproj` each third-party `PackageReference` has a both `PrivateAssets="all"` and a Version Range.

```xml
    <PackageReference Include="Microsoft.Extensions.Hosting"
                      Version="[3.1.3, 4.0.0)"
                      PrivateAssets="all"/>
```

- PrivateAssets forces consumers to reference `M.E.H` instead of having it transitively imported in their project.
 - The Version Range is nasty - what version am I really building against? That was my attempt to provide bounds for the supported versions in the nupkg definition.

Some guidance on how to approach this in a sustenable manner would be immensely beneficial. 
Thanks!

**Comment by tarekgh** [comment]
@shirhatti this issue has the information for guidance. We recommend using the version 5.0+ of the libraries. Users using older versions may need to consider migrating to 5.0 for better support (and we may need to find the data for the old breaking changes done in 2.1). Where can this guideline live? is it something you can help documenting it? should we move this issue to the doc repo? Thanks!

**Comment by shirhatti** [comment]
Before we move this to docs, we should be sure we don't plan to any product work here.

As an example, we could add an analyzer for library authors to avoid depending on public API that we broke? I'd like to see how much API surface we broke across all of Microsoft.Extensions.* before making this decision as this isn't just a logging concern

**Comment by tarekgh** [comment]
My understanding is in 5.0 we didn't introduce any break like the one introduced in 2.1. Does your ask concern all extension versions? or are you talking about moving from 3.1 to 5.0 breaking?

**Comment by eerhardt** [comment]
@maryamariyan @tarekgh - what are the next steps here?

I don't think we need to write an analyzer to ensure old APIs aren't being used. The old versions are now EOL (2.1 is no longer supported). We won't be making breaking changes in the M.E.* libraries going forward, so we don't need to worry about this in the future.

I believe our guidance for these packages remains the same as guidance for any NuGet dependency - so if some documentation needs to be written, it is broader than Microsoft.Extensions.*.

Basically, I believe this issue can be closed.

**Comment by tarekgh** [comment]
> 2.1 will no longer be supported

I don't think this is true for asp.net because of ASP.NET Core 2.1 on .NET Fx :-( 

@shirhatti offline mentioned `The guidance can be simplified to rely on the oldest supported Abstractions package if you are Core only. And there's no good solution if you're on .NET Fx`
